File: test/mock/user_test.go
package integration

import (
	"fmt"
	"testing"

	"github.com/ethereum/go-ethereum/common"
	"github.com/stretchr/testify/suite"
)

func TestUserSuite(t *testing.T) {
	suite.Run(t, new(UserSuite))
}

type UserSuite struct {
	DCMRollupSuite
}

func (s *UserSuite) TestCreateUser() {
	admin, _, creator, _, _, _, _, _ := s.setupCommonAddresses()
	baseTime, _, _ := s.setupTimeValues()

	// create creator user
	createUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"creator"}}`, creator))
	createUserOutput := s.Tester.Advance(admin, createUserInput)
	s.Len(createUserOutput.Notices, 1)

	expectedCreateUserOutput := fmt.Sprintf(`user created - {"id":3,"role":"creator","address":"%s","social_accounts":[],"created_at":%d}`, creator, baseTime)
	s.Equal(expectedCreateUserOutput, string(createUserOutput.Notices[0].Payload))
}

func (s *UserSuite) TestCreateInvestorUser() {
	admin, _, _, _, _, _, _, _ := s.setupCommonAddresses()
	investor01, _, _, _, _ := s.setupInvestorAddresses()
	baseTime, _, _ := s.setupTimeValues()

	// create investor user
	createUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"investor"}}`, investor01))
	createUserOutput := s.Tester.Advance(admin, createUserInput)
	s.Len(createUserOutput.Notices, 1)

	expectedCreateUserOutput := fmt.Sprintf(`user created - {"id":3,"role":"investor","address":"%s","social_accounts":[],"created_at":%d}`, investor01, baseTime)
	s.Equal(expectedCreateUserOutput, string(createUserOutput.Notices[0].Payload))
}

func (s *UserSuite) TestFindAllUsers() {
	admin, _, creator, _, _, _, _, _ := s.setupCommonAddresses()
	investor01, _, _, _, _ := s.setupInvestorAddresses()
	baseTime, _, _ := s.setupTimeValues()

	// create creator user
	createUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"creator"}}`, creator))
	s.Tester.Advance(admin, createUserInput)

	// create investor user
	createUserInput = []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"investor"}}`, investor01))
	s.Tester.Advance(admin, createUserInput)

	// find all users
	findAllUsersInput := []byte(`{"path":"user"}`)
	findAllUsersOutput := s.Tester.Inspect(findAllUsersInput)
	s.Len(findAllUsersOutput.Reports, 1)

	expectedFindAllUsersOutput := fmt.Sprintf(`[{"id":1,"role":"admin","address":"%s","social_accounts":[],"created_at":%d,"updated_at":0},{"id":2,"role":"verifier","address":"%s","social_accounts":[],"created_at":%d,"updated_at":0},{"id":3,"role":"creator","address":"%s","social_accounts":[],"created_at":%d,"updated_at":0},{"id":4,"role":"investor","address":"%s","social_accounts":[],"created_at":%d,"updated_at":0}]`,
		admin,
		baseTime,
		common.HexToAddress("0x0000000000000000000000000000000000000025"),
		baseTime,
		creator,
		baseTime,
		investor01,
		baseTime)
	s.Equal(expectedFindAllUsersOutput, string(findAllUsersOutput.Reports[0].Payload))
}

func (s *UserSuite) TestFindUserByAddress() {
	admin, _, creator, _, _, _, _, _ := s.setupCommonAddresses()
	baseTime, _, _ := s.setupTimeValues()

	// create creator user
	createUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"creator"}}`, creator))
	s.Tester.Advance(admin, createUserInput)

	// find user by address
	findUserByAddressInput := []byte(fmt.Sprintf(`{"path":"user/address","data":{"address":"%s"}}`, creator))
	findUserByAddressOutput := s.Tester.Inspect(findUserByAddressInput)
	s.Len(findUserByAddressOutput.Reports, 1)

	expectedFindUserByAddressOutput := fmt.Sprintf(`{"id":3,"role":"creator","address":"%s","social_accounts":[],"created_at":%d,"updated_at":0}`,
		creator,
		baseTime)
	s.Equal(expectedFindUserByAddressOutput, string(findUserByAddressOutput.Reports[0].Payload))
}

func (s *UserSuite) TestDeleteUser() {
	admin, _, _, _, _, _, _, _ := s.setupCommonAddresses()
	investor01, _, _, _, _ := s.setupInvestorAddresses()

	// create investor user
	createUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/create","data":{"address":"%s","role":"investor"}}`, investor01))
	s.Tester.Advance(admin, createUserInput)

	// delete user
	deleteUserInput := []byte(fmt.Sprintf(`{"path":"user/admin/delete","data":{"address":"%s"}}`, investor01))
	deleteUserOutput := s.Tester.Advance(admin, deleteUserInput)
	s.Len(deleteUserOutput.Notices, 1)

	expectedDeleteUserOutput := fmt.Sprintf(`user deleted - {"address":"%s"}`, investor01)
	s.Equal(expectedDeleteUserOutput, string(deleteUserOutput.Notices[0].Payload))
}

File: test/integration/user.test.ts
import { bytesToHex, stringToHex } from "viem";
import { afterAll, describe, expect, it } from "vitest";
import { encodeAdvanceInput, encodeNoticeOutput } from "./encoder";
import {
  createMachine,
  ADMIN_ADDRESS,
  VERIFIER_ADDRESS,
  CREATOR_ADDRESS,
  INVESTOR_01_ADDRESS,
} from "./helpers";

describe("User Tests", () => {
  const machine = createMachine();

  it("should create creator user", () => {
    const baseTime = Math.floor(Date.now() / 1000);

    const createUserInput = JSON.stringify({
      path: "user/admin/create",
      data: {
        address: CREATOR_ADDRESS,
        role: "creator",
      },
    });

    const { outputs } = machine.advance(
      encodeAdvanceInput({
        msgSender: ADMIN_ADDRESS,
        blockTimestamp: BigInt(baseTime),
        payload: `0x${Buffer.from(createUserInput).toString("hex")}`,
      }),
      { collect: true },
    );

    expect(outputs.length).toBe(1);

    const expectedNoticePayload = `user created - {"id":3,"role":"creator","address":"${CREATOR_ADDRESS}","social_accounts":[],"created_at":${baseTime}}`;
    const expectedOutput = encodeNoticeOutput({
      payload: stringToHex(expectedNoticePayload),
    });
    expect(bytesToHex(outputs[0])).toBe(expectedOutput);
  });

  it("should create investor user", () => {
    const baseTime = Math.floor(Date.now() / 1000);

    const createUserInput = JSON.stringify({
      path: "user/admin/create",
      data: {
        address: INVESTOR_01_ADDRESS,
        role: "investor",
      },
    });

    const { outputs } = machine.advance(
      encodeAdvanceInput({
        msgSender: ADMIN_ADDRESS,
        blockTimestamp: BigInt(baseTime),
        payload: `0x${Buffer.from(createUserInput).toString("hex")}`,
      }),
      { collect: true },
    );

    expect(outputs.length).toBe(1);

    const expectedNoticePayload = `user created - {"id":4,"role":"investor","address":"${INVESTOR_01_ADDRESS}","social_accounts":[],"created_at":${baseTime}}`;
    const expectedOutput = encodeNoticeOutput({
      payload: stringToHex(expectedNoticePayload),
    });
    expect(bytesToHex(outputs[0])).toBe(expectedOutput);
  });

  it("should find all users", () => {
    const baseTime = Math.floor(Date.now() / 1000);

    const findAllUsersInput = JSON.stringify({
      path: "user",
    });

    const reports = machine.inspect(Buffer.from(findAllUsersInput), {
      collect: true,
    });

    expect(reports.length).toBe(1);
    const output = JSON.parse(Buffer.from(reports[0]).toString("utf-8"));
    const expectedOutput = [
      {
        id: 1,
        role: "admin",
        address: ADMIN_ADDRESS,
        social_accounts: [],
        created_at: 0,
        updated_at: 0,
      },
      {
        id: 2,
        role: "verifier",
        address: VERIFIER_ADDRESS,
        social_accounts: [],
        created_at: 0,
        updated_at: 0,
      },
      {
        id: 3,
        role: "creator",
        address: CREATOR_ADDRESS,
        social_accounts: [],
        created_at: baseTime,
        updated_at: 0,
      },
      {
        id: 4,
        role: "investor",
        address: INVESTOR_01_ADDRESS,
        social_accounts: [],
        created_at: baseTime,
        updated_at: 0,
      },
    ];
    expect(output).toEqual(expectedOutput);
  });

  it("should find user by address", () => {
    const baseTime = Math.floor(Date.now() / 1000);

    const findUserByAddressInput = JSON.stringify({
      path: "user/address",
      data: {
        address: CREATOR_ADDRESS,
      },
    });

    const reports = machine.inspect(Buffer.from(findUserByAddressInput), {
      collect: true,
    });

    expect(reports.length).toBe(1);
    const output = JSON.parse(Buffer.from(reports[0]).toString("utf-8"));
    const expectedOutput = {
      id: 3,
      role: "creator",
      address: CREATOR_ADDRESS,
      social_accounts: [],
      created_at: baseTime,
      updated_at: 0,
    };
    expect(output).toEqual(expectedOutput);
  });

  it("should delete user", () => {
    const baseTime = Math.floor(Date.now() / 1000);

    const deleteUserInput = JSON.stringify({
      path: "user/admin/delete",
      data: {
        address: INVESTOR_01_ADDRESS,
      },
    });

    const { outputs } = machine.advance(
      encodeAdvanceInput({
        msgSender: ADMIN_ADDRESS,
        blockTimestamp: BigInt(baseTime),
        payload: `0x${Buffer.from(deleteUserInput).toString("hex")}`,
      }),
      { collect: true },
    );

    expect(outputs.length).toBe(1);

    const expectedNoticePayload = `user deleted - {"address":"${INVESTOR_01_ADDRESS}"}`;
    const expectedOutput = encodeNoticeOutput({
      payload: stringToHex(expectedNoticePayload),
    });
    expect(bytesToHex(outputs[0])).toBe(expectedOutput);
  });

  afterAll(() => {
    machine.shutdown();
  });
});
